### PR TITLE
fixes #19863; move sha1, md5 to nimble packages for 2.0

### LINKED ISSUE #19863: Replace SHA-1 and MD5 with non-deprecated hashing functions
### Summary
Both the SHA-1 and MD5 methods are deprecated, and should no longer be used for hashing.

### Description

The sha1 and md5 std modules in Nim should be deprecated (because those hashing methods themselves are deprecated) and replaced with sha2 and/or other newer hashing methods instead 

### Additional Information

Nim hashing modules: https://nim-lang.org/docs/lib.html#pure-libraries-hashing
Nim MD5 hashing: https://nim-lang.org/docs/md5.html
Nim SHA-1 hashing: https://nim-lang.org/docs/sha1.html

### PR DESCRIPTION
**Pull request by ringabout**
fixes #19863

ref https://github.com/nim-lang/checksums

- [x] documentation
- [x] move tests
- [x] add checksums to important packages
- [x] https://github.com/nim-lang/Nim/pull/21624

### CODE REVIEW

**Comment by ringabout** [comment]
I need to add a define so that nimble can work with this change, like `nimHasChecksum`.

**Comment by arnetheduck** [comment]
the alternative to moving them completely is to keep a deprecated version in stdlib for a release or two where people can keep using it but get a warning - then when they switch over, they get the same library without the warning etc

**Comment by ringabout** [comment]
> the alternative to moving them completely is to keep a deprecated version in stdlib for a release or two where people can keep using it but get a warning - then when they switch over, they get the same library without the warning etc

Good idea. I will consider it if it breaks too many important packages.

File: koch.nim
Comment (ringabout) [inline koch.nim:1]: clone the checksums

**Comment by ringabout** [comment]
Ci is broken; see also https://github.com/nim-lang/Nim/pull/21732

**Comment by tersec** [comment]
> the alternative to moving them completely is to keep a deprecated version in stdlib for a release or two where people can keep using it but get a warning - then when they switch over, they get the same library without the warning etc

There are things "deprecated" for years in Nim which haven't actually been removed. It would be good to have a pre-commited schedule to finally remove said modules if it proves necessary to delay things.

**Comment by c-blake** [comment]
The deprecation message could actually advertise at what version something will be removed.  Then client code can see that message and add tests for that with NimMajor/etc. instead of/as well as `when declared` type tests.  This might provide also added incentive to stick to the advertised schedule (which can vary by thing).  It would also make it easy to search for what should be removed as a Nim version release starts to be imminent.

**Comment by metagn** [comment]
Well the dominant metric for removing deprecations is time, and we don't really know when new versions release. Unless we scheduled versions too

**Comment by c-blake** [comment]
I do not mean scheduled in real-time but in release-time which solves that problem I think you refer to.

An example to be concrete: "This proc is deprecated to be removed in Nim-2.0; Use package FOO".  Or a Minor or patch number like "in Nim 2.1.2".  Something that just tells the person seeing the deprecation warning a marker they can then use in a `when (NimMajor,NimMinor,NimPatch) > (x,y,z)` condition.  Then whenever in real-time the release does drop things are all set.  A dev just needs a *little* lead time (to see the warning & react).

If the project had a long-term track record of doing this and sticking to it, that would make for evolution predictable enough that it would just be a habit of "see deprecation, go to nimble, put in guard around some `import`" - then it would work as tested on old & new and not require old-compiler users to import needlessly (with all that entails).

A similar `when` guard in a `.nimble` file for the dependency might be too much to ask to work (or maybe it does.  Not sure.  I'm no nimble expert.)  It feels there's a clean methodology to be had here.

**Comment by metagn** [comment]
Well the `(NimMajor, NimMinor)` check would be for the version it was deprecated, not the version it was removed, because we wouldn't want the deprecation message. Unless you are saying this is the version we should be mentioning, which was not what I understood. Maybe for standard library changes we can remove them as early as the next version, but for other things it can be hard to estimate how many packages need to update and how much effort and time they will each need to update. We could delay the scheduled removal date later on though, i.e. "Deprecated to be removed by 2.1.2 at the earliest"

(Also, in some cases like `smtp`, the code in the Nimble package still works for old versions like 0.15 and you don't need to use `when`, you can just unconditionally require the Nimble package. Might not be the case here)

**Comment by c-blake** [comment]
I don't think we are truly disagreeing, but you made me think I needed to be more clear.  So, here is an example from python-3.11.3 (you probably have to scroll to the right):
```python
import cgi
<stdin>:1: DeprecationWarning: 'cgi' is deprecated and slated for removal in Python 3.13
```
The version being mentioned is the one targeted for removal.  You do need/want the deprecation message to let people see it/react to it, though firebrands can always turn it off :-) .  There the "slated to" language is an escape hatch to shift later if necessary, like 3.14.

It's basically, "We promise you have until version X to adapt".  But yes, screaming protests from the crowd might get you to X+1 or +2 (or even a complete reversal of deprecation, I suppose).

**Comment by c-blake** [comment]
..oh, and sure, you *could* unconditionally require the nimble package.  I was mentioning something maybe slightly more nice or at least more precise - requiring only it for newer than certain Nim's (or in general any dependency, I suppose).  Maybe it's not more nice.  That's harder to say/more subjective.

**Comment by arnetheduck** [comment]
> It's basically, "We promise you have until version X to adapt". But yes, screaming protests from the crowd might get you to X+1 or +2 (or even a complete reversal of deprecation, I suppose).

this is what we're after: a target that people can plan their upgrades around. 

Upgrading Nim is complex and fraught with perils - it is often the case that it is not possible to upgrade between nim major releases until several point releases have passed to fix bugs - an overlap period _lessens_ the impact of changes and allows projects to stagger the upgrade: first upgrade nim itself (to get to a stable point with the new version), then deal with deprecations (so as to avoid a single massive upgrade). 

"slated to be deprecated" is indeed an excellent formulation and per the usual nim timelines, `X.(Y+1)` is likely a reasonable, sufficient and not-infinite overlap.

**Comment by arnetheduck** [comment]
> dominant metric for removing deprecations deprecations is time, 

I'd say it's releases since this is the only thing that the project controls and can make promises around - time flows on its own...

**Comment by Araq** [comment]
I largely agree but want to add to the discussion that a "deprecated yet never removed" symbol/feature still serves the major purpose of discouraging newly written code from using it. It's not a bad thing and could be a policy as well.

File: tests/manyloc/keineschweine/enet_server/server_utils.nim [hunk 1]
```diff
@@ -1,4 +1,6 @@
-import enet, sg_packets, estreams, md5, zlib_helpers, client_helpers, strutils,
+import ../../../../dist/checksums/src/checksums/md5
```
Comment (metagn) [inline tests/manyloc/keineschweine/enet_server/server_utils.nim:1]: Could also add `checksums` [here](https://github.com/nim-lang/Nim/blob/8e0f336f6dab061bd4dd0a07f4b436bbbe78138d/koch.nim#L559) and `import checksums/md5` in tests instead

**Pull request merged**